### PR TITLE
[JENKINS-51326] - Permit primitive types in the ClassFilterImpl

### PR DESCRIPTION
**Pull request by oleg-nenashev**
See [JENKINS-51326](https://issues.jenkins-ci.org/browse/JENKINS-51326). I am not sure why it happens in the particular reported case, because primitive classes should be boxed/unboxed by XStream. Obviously it does not happen in the reporter's case (maybe it's not XStream at all), to be investigated.

There is no sense to reject primitive classes, so I think this patch is reasonable anyway

### Proposed changelog entries

* RFE: Permit primitive types in the standard Jenkins Class Filter (JEP-200)
* ...

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees @jglick @daniel-beck

### CODE REVIEW

**Review by jglick — COMMENTED** [review]
Harmless but not a substitute for tracking down how primitive classes got passed here to begin with. It cannot be from Remoting, since that would only call it from `ObjectInputStreamEx.resolveClass`, which applies only to reference types. And clearly it does not normally happen from XStream, or we would see this constantly.

File: test/src/test/java/jenkins/security/ClassFilterImplTest.java
Comment (jeffret-b) [inline test/src/test/java/jenkins/security/ClassFilterImplTest.java]: I don't understand this comment. It seems to imply a check for string, but I don't see one.

**Review by jeffret-b — APPROVED** [review]
The changelog entry seems a little confusing: "Bug: Permit primitive types in the standard Jenkins Class Filter". When I read these sorts of things, I often wonder whether it describes the bug to be fixed or the desired behavior. (Maybe there is a standard pattern for this in Jenkins that I'm just not familiar with.)

Is this really addressing a bug?

File: test/src/test/java/jenkins/security/ClassFilterImplTest.java
Comment (oleg-nenashev) [reply test/src/test/java/jenkins/security/ClassFilterImplTest.java]: I just meant that the invocations below do nothing and hence they pass independently of what you pass

**Comment by oleg-nenashev** [comment]
@jeffret-b It would be displayed as `<red dot> Permit primitive types in the standard Jenkins Class Filter` in the changelog. Generally we may rather want to use it as RFE

File: test/src/test/java/jenkins/security/ClassFilterImplTest.java
Comment (jeffret-b) [reply test/src/test/java/jenkins/security/ClassFilterImplTest.java]: Ah, okay. i didn't read it closely enough.

**Comment by oleg-nenashev** [comment]
> Harmless but not a substitute for tracking down how primitive classes got passed here to begin with.

I have created #3447 . Once we have it, we will understand from were it comes at least

**Comment by dwnusbaum** [comment]
> I have created #3447 . Once we have it, we will understand from were it comes at least

Do we need to put this on hold until we diagnose the issues using #3447? Otherwise there will be no stack traces since the types will pass the whitelist, right?

**Comment by oleg-nenashev** [comment]
I would rather expect the reporter to run a core snapshot from here and report it before we merge anything

**Comment by oleg-nenashev** [comment]
On hold according to comments

**Comment by oleg-nenashev** [comment]
@jglick changed entry to RFE according to your feedback

**Review by jglick — CHANGES_REQUESTED** [review]
Hold on, there is new information in JIRA. Apparently [this code using `@XStreamAsAttribute`](https://github.com/jenkinsci/hockeyapp-plugin/blob/dfcaab51db811e29fdaad3c8183369d3e7721e97/src/main/java/hockeyapp/HockeyappApplication.java#L33-L35) leads to

```xml
<hockeyapp.HockeyappApplication plugin="hockeyapp@1.2.1" schemaVersion="1">
```

which is a format I have never seen before. If true, that means that we have a lead for writing an actual functional test reproducing the bug. I doubt the proposed fix is actually the right one—rather, `BlacklistedTypesConverter.canConvert` should be doing the `isPrimitive` check, or perhaps there is a bug even earlier in XStream somewhere, since the converter is not consulted for the regular child element format (is there a reason that would be autoboxed while attributes are not?).

File: test/src/test/java/jenkins/security/ClassFilterImplTest.java
Comment (dongsik2026) [inline test/src/test/java/jenkins/security/ClassFilterImplTest.java]: method name is good!

File: core/src/main/java/jenkins/security/ClassFilterImpl.java
Comment (dongsik2026) [inline core/src/main/java/jenkins/security/ClassFilterImpl.java]: I can't understand "c".

**Comment by oleg-nenashev** [comment]
@jglick I have no capacity to investigate it deeper and to fix XStream/whatever.
My proposal would be to merge the PR as a precaution and support for other potential use-cases (e.g. if other serializers want to support classfilters). If you and @jenkinsci/code-reviewers are against such approach, I would prefer to close the PR and unassign the ticket

**Comment by oleg-nenashev** [comment]
@jglick Ping, waiting for your response

**Review by jglick — CHANGES_REQUESTED** [review]
I do not think we should merge a fix without a (meaningful) test case and without even understanding the bug.

**Comment by oleg-nenashev** [comment]
Closing the PR since there is still no reproducible scenario from the reporter. I still believe this change could benefit from merging, but it is not the battle I want to spend time on

**Pull request closed without merging**
